### ISSUE TITLE
High cardinality for stacks_node_rpc_call_latencies_histogram metrics

### ISSUE THREAD

**Issue by CharlieC3** [stacks-network/stacks-core#4574]
With the latest release ([2.4.0.1.0](https://github.com/stacks-network/stacks-core/releases/tag/2.4.0.1.0)), the `stacks_node_rpc_call_latencies_histogram.*` prometheus metrics seemed to have changed and are now creating large amounts of cardinality due to the unique `path` labels being applied to them.
This can cause prometheus to consume more memory and become unusable pretty quickly. A short-term fix for this would be to drop the metric series from being scraped like so:
```
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: my-service-mon
spec:
  endpoints:
  - path: /metrics
    port: my-metrics-port
    relabelings:
    - action: drop
      regex: stacks_node_rpc_call_latencies_histogram.*
      sourceLabels:
      - __name__
...
```

However, a long-term fix would require making the `path` label less unique. The Stacks Blockchain API does something like this by globbing similar paths like so:
`/v2/attachments/012c8195f4381d0b0eadc2daee6e2c5b8243f4ff` would become `/v2/attachments/:attachment`

Very small snippet of the problematic metrics series (there are hundreds of thousands of these):
```
stacks_node_rpc_call_latencies_histogram_count{path="/v2/attachments/012c8195f4381d0b0eadc2daee6e2c5b8243f4ff"} 2
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="0.005"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="0.01"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="0.025"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="0.05"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="0.1"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="0.25"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="0.5"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="1"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="2.5"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="5"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="10"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a",le="+Inf"} 1
stacks_node_rpc_call_latencies_histogram_sum{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a"} 0.000895564
stacks_node_rpc_call_latencies_histogram_count{path="/v2/attachments/01586bc1be60143421cc3d7828e50bb26ded049a"} 1
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/016a3cb605d341b16c214dbb78b4bb90618bee80",le="0.005"} 2
stacks_node_rpc_call_latencies_histogram_bucket{path="/v2/attachments/016a3cb605d341b16c214dbb78b4bb90618bee80",le="0.01"} 2
```

**Comment by wileyj** [comment]
@jbencin any questions about the idea here, ping me or @CharlieC3

**Comment by jbencin** [comment]
Some questions:

- Which endpoints are affected by this, just `/v2/attachments`?
- Should I still keep track of the attachment, like `{ path="/v2/attachments/:attachment", attachment="012c81..." }`, or does that cause the same issue?
- This is in [2.4.0.1.0](https://github.com/stacks-network/stacks-core/releases/tag/2.4.0.1.0), so I should make my PR against `master`?

**Comment by wileyj** [comment]
> Some questions:
> 
> * Which endpoints are affected by this, just `/v2/attachments`?

@CharlieC3 any others you can think of?

> * Should I still keep track of the attachment, like `{ path="/v2/attachments/:attachment", attachment="012c81..." }`, or does that cause the same issue?

This is what causes the issue - the `:attachment` value will be different, so the system will start keeping track of ALL of them as individual metric buckets. this increases the memory used (marginally, but still) but more importantly it takes much longer to retrieve all the data (and the storage needs to store data for each attachment endpoint). 
for the API, what was done was basically to aggregate endpoints like this into a single bucket, i.e. `/v2/attachments` and the prom metrics returned data for all endpoints like `/v2/attachments/**`

> * This is in [2.4.0.1.0](https://github.com/stacks-network/stacks-core/releases/tag/2.4.0.1.0), so I should make my PR against `master`?

I think `next` is fine  - we likely won't have a release before 2.5 is merged to master

**Comment by CharlieC3** [comment]
> @CharlieC3 any others you can think of?

There are probably others, but I'm not sure. There were tens of thousands of attachment endpoints, I couldn't even load the `/metrics` endpoint to see any others lol, this might take a few tries. But generally any _generative_ endpoint which is not fully hardcoded should be abstracted away. So if an endpoint contains a transaction ID, or block ID or number, those should all be abstracted away as well. Non-generative hard-coded endpoints like `/v2/pox` or `/v2/neighbors` are good examples of safe endpoints which can be left alone.

> Yup `next` is fine for this issue since we have a small workaround in place.

**Comment by wileyj** [comment]
I'll take a look at the rpc endpoints and see ifi can identify any others
in the meantime

**Comment by wileyj** [comment]
only other endpoint i can find that matches the pattern @CharlieC3 shared is:
[/v2/blocks](https://github.com/stacks-network/stacks-core/blob/9b377f9c36357d6bdc7df4134c0bfd358c42c651/stackslib/src/net/api/getblock.rs#L90)

there are some GET endpoints that fall into the `microblocks` bucket, but i'm not sure it would be worth digging into that since the volume of requests there is likely to be extremely low.

**Comment by CharlieC3** [comment]
I would prefer we account for all dynamic endpoints if possible, otherwise it could be a potential vulnerability to indirectly affect observational infrastructure.

**Comment by CharlieC3** [comment]
Also, I believe this issue is causing wide-spread problems in metric collection. If a Prometheus server's `/metrics` page contains too many series to load due to high cardinality, the page may take over a minute to load or never load at all.
If an alert is configured to trigger in the absence of data scraped from a stacks node, or as an effect of outdated information from the last successful scrape, this will cause that to happen.

Restarting the node seems to temporarily help since the series' are flushed between each restart, but it's not a scalable solution.

**Comment by jbencin** [comment]
In the old HTTP stack there was a function `get_path()` which handled all the generated paths:

```rust
    pub fn get_path(&self) -> &'static str {
        match self {
            HttpRequestType::GetInfo(..) => "/v2/info",
            HttpRequestType::GetPoxInfo(..) => "/v2/pox",
            HttpRequestType::GetNeighbors(..) => "/v2/neighbors",
            HttpRequestType::GetHeaders(..) => "/v2/headers/:height",
            HttpRequestType::GetBlock(..) => "/v2/blocks/:hash",
            HttpRequestType::GetMicroblocksIndexed(..) => "/v2/microblocks/:hash",
            HttpRequestType::GetMicroblocksConfirmed(..) => "/v2/microblocks/confirmed/:hash",
            HttpRequestType::GetMicroblocksUnconfirmed(..) => {
                "/v2/microblocks/unconfirmed/:hash/:seq"
            }
            HttpRequestType::GetTransactionUnconfirmed(..) => "/v2/transactions/unconfirmed/:txid",
            HttpRequestType::PostTransaction(..) => "/v2/transactions",
            HttpRequestType::PostBlock(..) => "/v2/blocks/upload/:block",
            HttpRequestType::PostMicroblock(..) => "/v2/microblocks",
            HttpRequestType::GetAccount(..) => "/v2/accounts/:principal",
            HttpRequestType::GetDataVar(..) => "/v2/data_var/:principal/:contract_name/:var_name",
            HttpRequestType::GetMapEntry(..) => "/v2/map_entry/:principal/:contract_name/:map_name",
            HttpRequestType::GetTransferCost(..) => "/v2/fees/transfer",
            HttpRequestType::GetContractABI(..) => {
                "/v2/contracts/interface/:principal/:contract_name"
            }
            HttpRequestType::GetContractSrc(..) => "/v2/contracts/source/:principal/:contract_name",
            HttpRequestType::CallReadOnlyFunction(..) => {
                "/v2/contracts/call-read/:principal/:contract_name/:func_name"
            }
            HttpRequestType::GetAttachmentsInv(..) => "/v2/attachments/inv",
            HttpRequestType::GetAttachment(..) => "/v2/attachments/:hash",
            HttpRequestType::GetIsTraitImplemented(..) => "/v2/traits/:principal/:contract_name",
            HttpRequestType::MemPoolQuery(..) => "/v2/mempool/query",
            HttpRequestType::FeeRateEstimate(_, _, _) => "/v2/fees/transaction",
            HttpRequestType::OptionsPreflight(..) | HttpRequestType::ClientError(..) => "/",
        }
    }
```

I'm going to open a draft PR soon which addresses the more common endpoints. If necessary, we can address the rest later

**Comment by wileyj** [comment]
> In the old HTTP stack there was a function `get_path()` which handled all the generated paths:
> 
> ```rust
>     pub fn get_path(&self) -> &'static str {
>         match self {
>             HttpRequestType::GetInfo(..) => "/v2/info",
>             HttpRequestType::GetPoxInfo(..) => "/v2/pox",
>             HttpRequestType::GetNeighbors(..) => "/v2/neighbors",
>             HttpRequestType::GetHeaders(..) => "/v2/headers/:height",
>             HttpRequestType::GetBlock(..) => "/v2/blocks/:hash",
>             HttpRequestType::GetMicroblocksIndexed(..) => "/v2/microblocks/:hash",
>             HttpRequestType::GetMicroblocksConfirmed(..) => "/v2/microblocks/confirmed/:hash",
>             HttpRequestType::GetMicroblocksUnconfirmed(..) => {
>                 "/v2/microblocks/unconfirmed/:hash/:seq"
>             }
>             HttpRequestType::GetTransactionUnconfirmed(..) => "/v2/transactions/unconfirmed/:txid",
>             HttpRequestType::PostTransaction(..) => "/v2/transactions",
>             HttpRequestType::PostBlock(..) => "/v2/blocks/upload/:block",
>             HttpRequestType::PostMicroblock(..) => "/v2/microblocks",
>             HttpRequestType::GetAccount(..) => "/v2/accounts/:principal",
>             HttpRequestType::GetDataVar(..) => "/v2/data_var/:principal/:contract_name/:var_name",
>             HttpRequestType::GetMapEntry(..) => "/v2/map_entry/:principal/:contract_name/:map_name",
>             HttpRequestType::GetTransferCost(..) => "/v2/fees/transfer",
>             HttpRequestType::GetContractABI(..) => {
>                 "/v2/contracts/interface/:principal/:contract_name"
>             }
>             HttpRequestType::GetContractSrc(..) => "/v2/contracts/source/:principal/:contract_name",
>             HttpRequestType::CallReadOnlyFunction(..) => {
>                 "/v2/contracts/call-read/:principal/:contract_name/:func_name"
>             }
>             HttpRequestType::GetAttachmentsInv(..) => "/v2/attachments/inv",
>             HttpRequestType::GetAttachment(..) => "/v2/attachments/:hash",
>             HttpRequestType::GetIsTraitImplemented(..) => "/v2/traits/:principal/:contract_name",
>             HttpRequestType::MemPoolQuery(..) => "/v2/mempool/query",
>             HttpRequestType::FeeRateEstimate(_, _, _) => "/v2/fees/transaction",
>             HttpRequestType::OptionsPreflight(..) | HttpRequestType::ClientError(..) => "/",
>         }
>     }
> ```
> 
> I'm going to open a draft PR soon which addresses the more common endpoints. If necessary, we can address the rest later

good find! this is definitely a regression. 
@CharlieC3  can you clarify what you mean by `account for all dynamic endpoints if possible`? i understand that as any endpoint that may have a common root with a different tree node, i.e.
```
/v2/attachments/fake_hash_1
/v2/attachments/fake_hash_2
...
```

**Comment by CharlieC3** [comment]
> can you clarify what you mean by account for all dynamic endpoints if possible? i understand that as any endpoint that may have a common root with a different tree node

That is correct.

> I'm going to open a draft PR soon which addresses the more common endpoints. If necessary, we can address the rest later

@jbencin Any reason why address only some of the endpoints? Is it difficult to apply a fix for all generative endpoints? This problem can only be resolved once all generative endpoints are accounted for, as the symptoms will still be in effect otherwise. There are no endpoints in the stacks nodes Hiro hosts which receive little to no traffic due to bots and scrappers.

**Comment by wileyj** [comment]
> > can you clarify what you mean by account for all dynamic endpoints if possible? i understand that as any endpoint that may have a common root with a different tree node
> 
> That is correct.
> 
> > I'm going to open a draft PR soon which addresses the more common endpoints. If necessary, we can address the rest later
> 
> @jbencin Any reason why address only some of the endpoints? Is it difficult to apply a fix for all generative endpoints? This problem can only be resolved once all generative endpoints are accounted for, as the symptoms will still be in effect otherwise. There are no endpoints in the stacks nodes Hiro hosts which receive little to no traffic due to bots and scrappers.

are prom metrics being collected for POST only endpoints?

```
GET
/v2/headers/:height
/v2/blocks/:hash
/v2/microblocks/:hash
/v2/microblocks/confirmed/:hash
/v2/microblocks/unconfirmed/:hash/:seq
/v2/transactions/unconfirmed/:txid
/v2/accounts/:principal
/v2/data_var/:principal/:contract_name/:var_name
/v2/map_entry/:principal/:contract_name/:map_name
/v2/contracts/interface/:principal/:contract_name
/v2/contracts/source/:principal/:contract_name
/v2/contracts/call-read/:principal/:contract_name/:func_name
/v2/attachments/:hash
/v2/traits/:principal/:contract_name

POST
/v2/blocks/upload/:block
```

**Comment by jbencin** [comment]
I have a better PR for this now that fixes it for all endpoints: #4681

**Comment by CharlieC3** [comment]
Nice!! Thank you so much for looking into this 🙏
